README.md
# mobylette

mobylette is a python script written to parse system log files
and report which environment modules were loaded and how many times
each one was loaded. mobylette has a low memory footprint and can parse
several files simultaneously using several cpus, which is ideal for cluster
environments.

## tldr;

```
git clone https://github.com/scibian/mobylette
cd mobylette
source install.sh
mobylette
```
## Prerequisites

mobylette was designed to work with Lmod for Scibian 8 (since version 6.6-0.2sci8u2)
and Lmod for Scibian 9 (since version 6.6-0.2sci9u4).

## Configuration

mobylette uses a configuration file (`mobylete.conf`) which provides information about where to look for logs, which sub-directories to search and which files to read.

- `log_path` - mobylette will look for log files in this directory and all sub-directories.
- `patterns` - This field actually accepts regular expressions (python compatible). As an example, mobylette can search only compressed files if the following regex `filename[\S]gz` is used.

mobylette was designed to run in a cluster environment. As such, there can be hundreds or thousands of sub-directories where to look for logs. mobylette assumes these sub-directories have a naming mechanism which can be seen as `prefix` + `nodes`.

- `prefix` - Prefix for the nodes, usually the first letters of the cluster name.
- `nodes` - Usually a cluster has different type of nodes. Can accept multiple values separated by commas.
 
If both of these two fields are left empty, mobylette will search for files under the directory given by log_path and all its sub-directories.
 
mobylette will look for the configuration file in `.mobylette` under the user home directory and then in the `/etc` directory.

## Installation

After having cloned the repository the `install.sh` script should be sourced. This will update the `PATH` and `PYTHONPATH` variables with the location of mobylette files.

In case it is the first time you are using mobylette, this script will also create a dummy configuration file under `.mobylette` in the users home directory. This out-of-the-box configuration sets up mobylette to run against two log files which are given in the sample directory.

## Getting started

The `mobylette` command will create a csv file (in the same directory where the command was issued) and also a horizontal bar graph relating the data found in the csv file.

With the out-of-the-box configuration (i.e., mobylette.conf points to the sample directory) the following data can be seen:

```
abc/123,1
gcc/9.2,3
ifaible/2020,1
ifaible/3030,2
ifort/2020,1
ifort/3030,2
python/2.7,2
python/3.6,1
```

![Number of module load commands](sample/0_abc123_python3.6.svg)


## Use cases

    $ mobylette

The standard behaviour is to count the number of modules found in distinct jobs. 

    $ mobylette -group cat
    
The `-group cat`option tells mobylette to group the results by module category. That is, the name of the module or the string that preceds the version of the module.

    $ mobylette -group path
    
When grouping by the modules path, only the first directory of the path is considered.

    $ mobylette -uniq users

Other than counting the number of module across different jobs, mobylette can also count the number of modules loaded by diferent users. The report file created with this option also lists the users that loaded the module.

    $ mobylette -uniq users -group cat

As before, modules can be grouped in categories, also when the different users paradigm is in use.

    $ mobylette -uniq users -group path
    
Grouping by path is also available when counting unique users.

## Filter options

    -start yyyymmdd

When start parameter is specified, mobylette will only retain system log entries whose date is AFTER the given parameter.
If desired, this parameter will also accept the time passed immediately after the date (no space between) in hhmmss format.

    -end yyyymmdd
    
If specified, mobylette will only retain system log entries whose date is BEFORE the given parameter.
If desired, this parameter will also accept the time passed immediately after the date (no space between) in hhmmss format.

    -module <module1> <module1>
    
A list of modules can be given to mobylette. As expected mobylete will check if the modules found in the logs belong to this list before updating the counts.
    
## Other options
    
    -cpus n
    
By default, mobylette will take up to 75% of the existing cores. This behaviour can be changed by explicitly giving the number od cores to be used with the `-cpus` parameter. No need to say, don't run mobylette with more cores than you have.

    -verbose
    
Can be usedfull, who knows?

    -chart-color <hex-value, no pound sign>
    
Faishon is always changing. Try to keep up!
    
## Authors

EDF CCN-HPC

## License

mobylette is distributed under the terms of the GPL v3 licence.


debian/changelog
mobylette (2.5-0sci9u1) scibian9; urgency=medium

  * New Scibian release

 -- Eduardo Rothe <eduardo-externe.rothe@edf.fr>  Tue, 05 Mar 2020 11:11:11 +0100


install.sh
#!/bin/bash

MOBYLETTE_HOME=`pwd`
export PATH=$MOBYLETTE_HOME/bin:$PATH
export PYTHONPATH=$MOBYLETTE_HOME:$PYTHONPATH

# Creates mobylette.conf
CONF_FILE=mobylette.conf
CONF_PATH=`eval echo ~$USER`/.mobylette

if [ -d "$CONF_PATH" ]; then
    read -p "$CONF_PATH already exists. Overwrite (y/n)? " ANSWER

    if [ "$ANSWER" = "n" ]; then
        echo "mobylette ready"
        return
    fi

    if [ "$ANSWER" = "y" ]; then

        rm -r $CONF_PATH
        mkdir $CONF_PATH

        echo "[CLUSTER]"                       > $CONF_PATH/$CONF_FILE
        echo "log_path=$MOBYLETTE_HOME/"      >> $CONF_PATH/$CONF_FILE
        echo "name=phoenix"                   >> $CONF_PATH/$CONF_FILE
        echo "nodes="                         >> $CONF_PATH/$CONF_FILE
        echo "patterns=dataset"               >> $CONF_PATH/$CONF_FILE
        echo "prefix=sample"                  >> $CONF_PATH/$CONF_FILE

        echo "mobylette ready"
        return
    fi
fi


mobylette/chart.py
# -*- coding: utf-8 -*-
##############################################################################
#                                                                            #
#  This file is part of the mobylette parsing tool.                          #
#        Copyright (C) 2019 EDF SA                                           #
#                                                                            #
#  mobylette is free software: you can redistribute it and/or modify         #
#  it under the terms of the GNU General Public License as published by      #
#  the Free Software Foundation, either version 3 of the License, or         #
#  (at your option) any later version.                                       #
#                                                                            #
#  mobylette is distributed in the hope that it will be useful,              #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of            #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the              #
#  GNU General Public License for more details.                              #
#                                                                            #
#  You should have received a copy of the GNU General Public License         #
#  along with mobyllette. If not, see <http://www.gnu.org/licenses/>.        #
#                                                                            #
##############################################################################
''' Implements Chart class '''

from matplotlib import use
use('Svg')
import matplotlib.pyplot as plt
plt.rcdefaults()
from matplotlib.ticker import MaxNLocator
from matplotlib.ticker import MultipleLocator

class Chart(object):

    def __init__(self, x, y, options, x_label, title, chart_color = None):
        self.options = options
        self.x_label = x_label
        self.title = title
        self.x = x
        self.y = y
        if chart_color is None:
            self.chart_color = '#2792ea'
        else:
            self.chart_color = '#' + chart_color
        self.setup_chart()

    def setup_chart(self):
        ''' Creates several DataFrames '''
        #import pdb ; pdb.set_trace()
        if self.options['chart']['max_charts'] is not None:
            setup = self._distribute_charts(len(self.x), self.options['chart']['max_charts'])
        else:
            setup = self._distribute_rows(len(self.x), self.options['chart']['max_rows'])
        boxes = self._seq_to_list(setup)
        for count, interval in enumerate(boxes):
            self.do_simple(self.x[interval[0]:interval[1] + 1], self.y[interval[0]:interval[1] + 1], count)

    def _seq_to_list(self, seq):
        ''' Returns list of tupples representing the intervals
        of 3 sets `len(seq)`, each one having the corresponding
        number of elements.

        Example:

        Given    :   [3,3,2],
        returns  :   [(0, 2), (3, 5), (6, 7)]
        '''
        result = []
        for count, item in enumerate(seq):
            floor = 0 if count == 0 else result[count - 1][1] + 1
            ceiling = floor + item - 1
            result.append((floor, ceiling))
        return(result)

    def _distribute_charts(self, balls, box_num):
        ''' Distributes balls over boxes.

        Returns : list with the number of balls per box,
                  such that the difference of number of
                  balls in two different boxes cannot be
                  greater than 1. That is, the distribution
                  of balls over boxes is equillibrated.

        Example : balls = 31, box_dim = 7
                  31 / 7 = 4
                  31 % 7 = 3
                  4 boxes will have 4 balls and 3 will have 5 balls.
                  4*4 + 3*5 = 16 + 15 = 31
        '''
        base = balls / box_num
        if base == 0:
            base = 1
        extra_ball_boxes = balls % box_num
        result = [base] * box_num
        while extra_ball_boxes > 0:
            result[extra_ball_boxes - 1] += 1
            extra_ball_boxes -= 1
        return(result)

    def _distribute_rows(self, balls, box_dim):
        ''' Distributes balls over boxes.

        Returns : list with the number of balls for each box,
                  knowing that each box can have no more than
                  box_dim + 1 balls.

        Example : balls = 31, box_dim = 7
                  31 / 7 = 4
                  31 % 7 = 3
                  4 boxes will have 4 balls and 3 will have 5 balls.
                  4*4 + 3*5 = 16 + 15 = 31
        '''

        box_num = balls / box_dim
        if box_num == 0:
            box_num = 1
        extra_balls = balls % box_dim
        result = [box_dim] * box_num
        for i in range(extra_balls):
            result[i%len(range(box_num))] += 1
        return(result)
        
    def do_simple(self, x, y, count):
        ''' Creates horizontal bar chart with ordered
        labels and stores the image in a SVG file whose
        name is defined by the first and last labels of
        the elements being represented. Count will also
        enter in the composition of the file name.
        '''
        L = sorted(zip(x,y), reverse=True)
        x = [i[0] for i in L]
        y = [i[1] for i in L]

        fig, ax = plt.subplots()
        plt.xlabel(self.x_label)
        ax.set_title(self.title)
        plt.gca().spines['right'].set_visible(False)
        plt.gca().spines['left'].set_visible(False)
        plt.tick_params(top=False, bottom=True, left=False, right=False, labelleft=True, labelbottom=True)
        ax.barh(range(len(x)), y, facecolor=self.chart_color, align='center', edgecolor='white')
        plt.yticks(range(len(x)), x)
        ax.margins(x=0, y=0.05)
        ax.xaxis.set_major_locator(MaxNLocator(integer=True))
        plt.locator_params(axis='x', nbins=5)
        file_name = str(count) + '_' + x[-1] + '_' + x[0] + '.svg'
        plt.savefig(file_name.replace('/', ''), bbox_inches='tight')


mobylette/config.py
# -*- coding: utf-8 -*-
##############################################################################
#                                                                            #
#  This file is part of the mobylette parsing tool.                          #
#        Copyright (C) 2019 EDF SA                                           #
#                                                                            #
#  mobylette is free software: you can redistribute it and/or modify         #
#  it under the terms of the GNU General Public License as published by      #
#  the Free Software Foundation, either version 3 of the License, or         #
#  (at your option) any later version.                                       #
#                                                                            #
#  mobylette is distributed in the hope that it will be useful,              #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of            #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the              #
#  GNU General Public License for more details.                              #
#                                                                            #
#  You should have received a copy of the GNU General Public License         #
#  along with mobyllette. If not, see <http://www.gnu.org/licenses/>.        #
#                                                                            #
##############################################################################
''' Implements Config class'''

import os
import sys
import pwd
import subprocess
if sys.version_info.major == 2:
    import ConfigParser as configparser
else:
    import configparser as configparser

class Config(object):
    '''
    Reads configuration from mobylette.conf which can be found in:

      /etc/mobylette.conf
      ~/.mobylette/mobylette.conf

    More search paths can be added to conf_path. The first path where
    mobylette.conf is found will be returned.
    '''
    
    conf_path = [u'' + pwd.getpwuid(os.getuid()).pw_dir + '/.mobylette/mobylette.conf',
                 u'' + '/etc/mobylette.conf']
    config = configparser.ConfigParser()

    def __init__(self, verbose):
        self.verbose = verbose
        self.log_path = None
        self.cluster_name = None
        self.cluster_prefix = None
        self.search_pattern = None
        self.nodes_tuple = None
        self.hostname = self._read_hostname()
        self.read_configuration()

    def _read_hostname(self):
        ''' Read hostname from command line.
        hostname read from command line Will be used to match against cluster name found in mobylette.conf
        '''
        process_result = subprocess.Popen('hostname', stdout=subprocess.PIPE, shell=True)
        (output, err) = process_result.communicate() 
        result_status = process_result.wait()
        if result_status != 0:
            print('Could not read cluster name : {}'.format(err.rstrip().decode('utf8')))
        hostname = output.rstrip().decode('utf8')
        if self.verbose:
            print('hostname: {}'.format(hostname))
        return hostname

    def _config_in_location(self, location):
        ''' Returns True if configparser is able to read a file in given location '''
        f = self.config.read(location)
        result = False
        if len(f) > 0:
            result = True
            if self.verbose:
                print('found configuration file at: {}'.format(location))
        return result

    def _config_file(self):
        ''' Returns a handle to the config file '''
        j = 0
        while not self._config_in_location(self.conf_path[j]):
            j = j + 1
        return self.config.read(self.conf_path[j])

    def read_configuration(self):
        ''' Reads config file '''
        f = self._config_file()
        if len(f) > 0:
            if self.config.has_option('CLUSTER', 'name'):
                self.cluster_name = self.config.get('CLUSTER', 'name')
            if self.config.has_option('CLUSTER', 'prefix'):
                self.cluster_prefix = self.config.get('CLUSTER', 'prefix')
            if self.config.has_option('CLUSTER', 'log_path'):
                self.log_path = self.config.get('CLUSTER', 'log_path')
            if self.config.has_option('CLUSTER', 'patterns'):
                self.search_pattern = self.config.get('CLUSTER', 'patterns')
            if self.config.has_option('CLUSTER', 'nodes'):
                # self.nodes_tuple = tuple([ self.hostname[0:2] + element for element in self.config.get('CLUSTER', 'nodes').split(',')])
                self.nodes_tuple = tuple([ self.cluster_prefix + element for element in self.config.get('CLUSTER', 'nodes').split(',')])


mobylette/__init__.py
##############################################################################
#                                                                            #
#  This file is part of the mobylette parsing tool.                          #
#        Copyright (C) 2019 EDF SA                                           #
#                                                                            #
#  mobylette is free software: you can redistribute it and/or modify         #
#  it under the terms of the GNU General Public License as published by      #
#  the Free Software Foundation, either version 3 of the License, or         #
#  (at your option) any later version.                                       #
#                                                                            #
#  mobylette is distributed in the hope that it will be useful,              #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of            #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the              #
#  GNU General Public License for more details.                              #
#                                                                            #
#  You should have received a copy of the GNU General Public License         #
#  along with mobyllette. If not, see <http://www.gnu.org/licenses/>.        #
#                                                                            #
##############################################################################


mobylette/args.py
# -*- coding: utf-8 -*-
##############################################################################
#                                                                            #
#  This file is part of the mobylette parsing tool.                          #
#        Copyright (C) 2019 EDF SA                                           #
#                                                                            #
#  mobylette is free software: you can redistribute it and/or modify         #
#  it under the terms of the GNU General Public License as published by      #
#  the Free Software Foundation, either version 3 of the License, or         #
#  (at your option) any later version.                                       #
#                                                                            #
#  mobylette is distributed in the hope that it will be useful,              #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of            #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the              #
#  GNU General Public License for more details.                              #
#                                                                            #
#  You should have received a copy of the GNU General Public License         #
#  along with mobyllette. If not, see <http://www.gnu.org/licenses/>.        #
#                                                                            #
##############################################################################
''' Implements ParseArgs class'''

import argparse                                     # ArgumentParser, RawTextHelpFormatter, SUPPRESS
from datetime import datetime as dt                 # fromtimestamp, now, strptime

class ParseArgs(object):
    ''' Reads arguments from command line

    The full path of a module is:

        mod path                 mod name
            |                       |
      /some_path/some_category/some_module/some_version
                     |                          |
                mod category               mod version

     mod version - this is the file, version of the module
     mod name - the directory where the `mod name` file exists. It is the name of the module
     mod category - is the directory where `mod name` directory exists (it could also be named library)
     mod path - is a path (more than one directory)
    '''

    parser = argparse.ArgumentParser(
                        formatter_class=argparse.RawTextHelpFormatter,
                        description="Parses system logs for Lmod `module load` entries and writes a csv report\n" +
                                    "with the result found as well as a beautiful SVG graph to tell the story!")

    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
    # *                                                             *
    # * COUNT                                                       *
    # *                                                             *
    # *                                                             *
    # *                                                             *
    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

    parser.add_argument('--version', action='version', version='mobylette 2.5')

    # What to count: users or modules
    parser.add_argument('-uniq', action='store', choices=['users', 'jobs'], dest='uniq', default=['jobs'],
                        help="" +
                             "Whether unique users or jobs are to be considered\n" +
                             "The default is to count modules found in distinct\n" +
                             "jobs. The -users option will create a list having\n" +
                             "the usernames of the users who loaded the module.")

    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
    # *                                                             *
    # * GROUP BY                                                    *
    # *                                                             *
    # *                                                             *
    # *                                                             *
    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

    # How counts will be gruoped by
    parser.add_argument('-group', action='store', choices=['cat', 'path'], dest='group',
                        help="" +
                             "Data may be grouped by the modules category or by\n" +
                             "the first directory of it's path. No option given\n" +
                             "no grouping is done.") 

    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
    # *                                                             *
    # * FILTER                                                      *
    # *                                                             *
    # *                                                             *
    # *                                                             *
    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

    # Start date
    parser.add_argument('-start', action='store', metavar="<yyyymmdd>", dest="start_date",
                        help="" +
                             "Modules have been loaded before this date will be\n" +
                             "be ignored. If no argument is passed then all the\n" +
                             "modules will be read.")

    # End date
    parser.add_argument('-end', action='store', metavar="<yyyymmdd>", dest="end_date",
                        help="" +
                             "If a module was loaded after this date it will be\n" +
                             "ignored. The default behaviour is to read modules\n" +
                             "of all dates.")

    # Module list
    parser.add_argument('-module', action='store', nargs='+', metavar="<module>", dest="module",
                        help="" +
                             "Modules can go here. mobylette can receive a list\n" +
                             "of modules and will exclude all other modules not\n" +
                             "in this list. The default behaviour will consider\n" +
                             "all modules. White space separated list. Both the\n" +
                             "name and version of the module should be present.\n")


    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
    # *                                                             *
    # * Hidden options                                              *
    # *                                                             *
    # *                                                             *
    # *                                                             *
    # * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

    # Verbose
    parser.add_argument('-verbose', action='store_true', dest="verbose",
                        help=argparse.SUPPRESS)

    # Number of cpus to use while parsing the files
    parser.add_argument('-cpus', action='store', dest="cpus", type=int,
                        help=argparse.SUPPRESS)

    # Color fo the bars in the chart
    parser.add_argument('-chart-color', action='store', dest="chart_color",
                        help=argparse.SUPPRESS)

    # This is actually a trick to force namespace to create a parameter called 'pattern'
    # Later this parameter will be replaced
    parser.add_argument('-pattern', action='store', dest="pattern",
                        help=argparse.SUPPRESS)

    # The following two parameters control how to distribute the data over the charts
    # so that the chart does not become over crowded (and impossible to read).
    chart_size = parser.add_mutually_exclusive_group()
    chart_size.add_argument('-max-charts', action='store', dest="max_charts", type=int,
                        help=argparse.SUPPRESS)
    chart_size.add_argument('-max-rows', action='store', dest="max_rows", default=13, type=int,
                        help=argparse.SUPPRESS)

    args = parser.parse_args()

    def __init__(self):
        if 'figsize' not in self.args:
            self.args.figsize = None
        if 'dpi' not in self.args:
            self.args.dpi = None
        if 'cpus' not in self.args:
            self.args.cpus = None

        if self.args.start_date is not None:
            self.args.start_date = self.str2timestamp(self.args.start_date)
        if self.args.end_date is not None:
            self.args.end_date = self.str2timestamp(self.args.end_date)

        if 'pattern' not in self.args:
            self.args.pattern = None

        if 'max_rows' not in self.args:
            self.args.max_rows = None
        if 'max_charts' not in self.args:
            self.args.max_charts = None
            
    def str2date(self, date):
        '''
           input: yyyymmdd             output: datetime object
                  yyyymmddhhmmss
                  yyyymmddThhmmss
        '''
        if len(date) == 8:
            result = dt.strptime(date + '000000','%Y%m%d%H%M%S')
        elif len(date) == 14:
            result = dt.strptime(date, '%Y%m%d%H%M%S')
        elif len(date) == 15:
            result = dt.strptime(date, '%Y%m%dT%H%M%S')
        return(result)

    def str2timestamp(self, str_date):
        '''
        args:
          string: yyyymmdd             
                  yyyymmddhhmmss
                  yyyymmddThhmmss
        return:
          integer
        '''
        from time import mktime
        return(int(mktime(self.str2date(str_date).timetuple())))

    def date2timestamp(self, dt_obj):
        '''
        args:
          datetime.datetime

        return:
          integer
        '''
        from time import mktime
        return(int(mktime(dt_obj.timetuple())))
